### PR TITLE
Services/APT: Better implementation of PrepareToDoApplicationJump and DoApplicationJump.

### PR DESCRIPTION
**Pull request by Subv**
The real console can't launch an Application directly from within another Application so it has to go through the Home Menu. We do not have such limitation and can directly launch the requested title.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4306)

### CODE REVIEW

File: src/core/hle/service/apt/applet_manager.cpp [hunk 1]
```diff
@@ -465,6 +469,93 @@ ResultVal<AppletManager::AppletInfo> AppletManager::GetAppletInfo(AppletId app_i
...
+    // prompts it to call GetProgramIdOnApplicationJump and
+    // PrepareToStartApplication/StartApplication on the title to launch.
+
+    if (app_jump_parameters.next_title_id == app_jump_parameters.current_title_id) {
+        // Perform a soft-reset if we're trying to relaunch the same title.
+        // TODO(Subv): Note that this reboots the entire emulated system, a better way would be to
+        // simply re-launch the title without closing all services, but this would only work for
+        // installed titles since we have no way of getting the file path of an arbitrary game dump
+        // based only on the title id.
+        Core::System::GetInstance().RequestReset();
```
Comment (Subv) [inline src/core/hle/service/apt/applet_manager.cpp]: I don't have a way to access APT's `system` member from within the `AppletManager`, should I pass the System instance to the AppletManager constructor (or perhaps an APT Module instance)? @wwylele

Comment (wwylele) [reply src/core/hle/service/apt/applet_manager.cpp]: Yes, feel free to pass the system reference into AppletManager.

File: src/core/hle/service/apt/applet_manager.cpp [hunk 2]
```diff
@@ -465,7 +469,94 @@ ResultVal<AppletManager::AppletInfo> AppletManager::GetAppletInfo(AppletId app_i
...
+    // prompts it to call GetProgramIdOnApplicationJump and
+    // PrepareToStartApplication/StartApplication on the title to launch.
+
+    if (app_jump_parameters.next_title_id == app_jump_parameters.current_title_id) {
+        // Perform a soft-reset if we're trying to relaunch the same title.
+        // TODO(Subv): Note that this reboots the entire emulated system, a better way would be to
+        // simply re-launch the title without closing all services, but this would only work for
+        // installed titles since we have no way of getting the file path of an arbitrary game dump
+        // based only on the title id.
+        system.RequestReset();
```
[the remark this answers is not in the document]
Comment (Subv) [reply src/core/hle/service/apt/applet_manager.cpp]: I think only the loader knows of the file path and it's hidden behind several layers of abstraction. In the general case, we can't get the file path based only on the title id for non-installed titles.

Implementing that would be nice but it's out of scope for this PR.

[the remark this answers is not in the document]
Comment (zhaowenlan1779) [reply src/core/hle/service/apt/applet_manager.cpp]: iirc it's just a data member.

File: src/core/hle/service/apt/apt.cpp [hunk 3]
```diff
@@ -394,6 +394,54 @@ void Module::Interface::CancelParameter(Kernel::HLERequestContext& ctx) {
               static_cast<u32>(receiver_appid));
 }
 
+void Module::Interface::PrepareToDoApplicationJump(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x31, 4, 0); // 0x00310100
+    auto flags = rp.PopEnum<ApplicationJumpFlags>();
+    u64 title_id = rp.Pop<u64>();
+    u8 media_type = rp.Pop<u8>();
+
+    LOG_WARNING(Service_APT, "(STUBBED) called title_id={:016X}, media_type={:#01X}, flags={:#08X}",
```
Comment (B3n30) [inline src/core/hle/service/apt/apt.cpp]: LOG_DEBUG? Remove (STUBBED) ?

File: src/core/hle/service/apt/applet_manager.h [hunk 4]
```diff
@@ -169,19 +196,26 @@ class AppletManager : public std::enable_shared_from_this<AppletManager> {
...
+    ApplicationJumpParameters app_jump_parameters{};
+
     // Holds data about the concurrently running applets in the system.
     std::array<AppletSlotData, NumAppletSlot> applet_slots = {};
 
     // This overload returns nullptr if no applet with the specified id has been started.
     AppletSlotData* GetAppletSlotData(AppletId id);
     AppletSlotData* GetAppletSlotData(AppletAttributes attributes);
 
+    void EnsureHomeMenuLoaded();
```
Comment (B3n30) [inline src/core/hle/service/apt/applet_manager.h]: this isn't used anywhere

Comment (Subv) [reply src/core/hle/service/apt/applet_manager.h]: Yes, it's there as documentation for a future implementation. This is explained in the comments I added

File: src/core/hle/service/apt/apt.cpp (hunk 3 above)
Comment (Subv) [reply src/core/hle/service/apt/apt.cpp]: It's not a complete implementation, see `AppletManager::PrepareToDoApplicationJump` for details

**Pull request merged**